sympy/polys/numberfields/modules.py

```diff
@@ -0,0 +1,1794 @@
+    return DomainMatrix([[ZZ(c) for c in coeffs]], (1, len(coeffs)), ZZ).transpose()
+
+
+class Module:
+    """
+    Generic finitely-generated module.
+
+    This is an abstract base class, and should not be instantiated directly.
+    The two concrete subclasses are :py:class:`~.PowerBasis` and
+    :py:class:`~.Submodule`.
+
+    Every :py:class:`~.Submodule` is derived from another module, referenced
+    by its ``parent`` attribute. If ``S`` is a submodule, then we refer to
+    ``S.parent``, ``S.parent.parent``, and so on, as the "ancestors" of
+    ``S``. Thus, every :py:class:`~.Module` is either a
+    :py:class:`~.PowerBasis` or a :py:class:`~.Submodule`, some ancestor of
+    which is a :py:class:`~.PowerBasis`.
+    """
+
+    @property
+    def n(self):
+        """The number of generators of this module."""
+        raise NotImplementedError
+
+    def mult_tab(self):
+        """
+        Get the multiplication table for this module (if closed under mult).
+
+        Explanation
+        ===========
+
+        Computes a dictionary ``M`` of dictionaries of lists, representing the
+        upper triangular half of the multiplication table.
+
+        In other words, if ``0 <= i <= j < self.n``, then ``M[i][j]`` is the
+        list ``c`` of coefficients such that
+        ``g[i] * g[j] == sum(c[k]*g[k], k in range(self.n))``,
+        where ``g`` is the list of generators of this module.
+
+        If ``j < i`` then ``M[i][j]`` is undefined.
+
+        Examples
+        ========
+
+        >>> from sympy.polys import Poly, cyclotomic_poly
+        >>> from sympy.polys.numberfields.modules import PowerBasis
+        >>> T = Poly(cyclotomic_poly(5))
+        >>> A = PowerBasis(T)
+        >>> print(A.mult_tab())  # doctest: +SKIP
+        {0: {0: [1, 0, 0, 0], 1: [0, 1, 0, 0], 2: [0, 0, 1, 0],     3: [0, 0, 0, 1]},
+                          1: {1: [0, 0, 1, 0], 2: [0, 0, 0, 1],     3: [-1, -1, -1, -1]},
+                                           2: {2: [-1, -1, -1, -1], 3: [1, 0, 0, 0]},
+                                                                3: {3: [0, 1, 0, 0]}}
+
+        Returns
+        =======
+
+        dict of dict of lists
+
+        Raises
+        ======
+
+        ClosureFailure
+            If the module is not closed under multiplication.
+
+        """
+        raise NotImplementedError
+
+    @property
+    def parent(self):
+        """
+        The parent module, if any, for this module.
+
+        Explanation
+        ===========
+
+        For a :py:class:`~.Submodule` this is its ``parent`` attribute; for a
+        :py:class:`~.PowerBasis` this is ``None``.
+
+        Returns
+        =======
+
+        :py:class:`~.Module`, ``None``
+
+        See Also
+        ========
+
+        Module
+
+        """
+        return None
+
+    def represent(self, elt):
+        """
+        Represent an element of an ancestor module as an integer-linear
+        combination over the generators of this module.
```

Comment:
Maybe this docstring should be the one that actually explains what `represent` means rather than the two subclasses. I think it would be good to have (somewhere) a demonstration of how represent works showing how a particular number is represented in a particular module and how you could reconstruct the `Expr` form of a number from the generators and the vector returned here by represent.